src/cargo/sources/git/utils.rs

```diff
@@ -217,12 +333,39 @@ impl GitReference {
                     .target()
                     .ok_or_else(|| anyhow::format_err!("branch `{}` did not have a target", s))?
             }
+
+            // See the module docs for why we're using `master` here.
             GitReference::DefaultBranch => {
-                let refname = "refs/remotes/origin/HEAD";
-                let id = repo.refname_to_id(refname)?;
-                let obj = repo.find_object(id, None)?;
-                let obj = obj.peel(ObjectType::Commit)?;
-                obj.id()
+                let master = repo
+                    .find_branch("origin/master", git2::BranchType::Remote)
+                    .chain_err(|| "failed to find branch `master`")?;
+                let master = master
+                    .get()
+                    .target()
+                    .ok_or_else(|| anyhow::format_err!("branch `master` did not have a target"))?;
+
+                if let Some((remote, config)) = remote_and_config {
+                    let head_id = repo.refname_to_id("refs/remotes/origin/HEAD")?;
+                    let head = repo.find_object(head_id, None)?;
+                    let head = head.peel(ObjectType::Commit)?.id();
+
+                    if head != master {
+                        config.shell().warn(&format!(
+                            "\
+                                fetching `master` branch from `{}` but the `HEAD` \
```

Comment:
This warning will fire in the case when someone has switched to the main branch since and the master branch is outdated, right? In that case, the suggestion might be wrong, maybe what the user wants is `branch = "main"`. Cargo can't tell the new default branch but hinting at the possibility in the warning would be great (technically, cargo could list all branches and filter the ones which match with HEAD but that would require one would have to fetch basically all branches which is probably not something you'd want to do :p).

Reply:
Yes the warning will fire in that case, but Cargo is not using `HEAD` and is instead using the dependency as-if `branch = "master"` was specified. The suggestion here is to write down in `Cargo.toml` how to perform the same build in the future. It's true though that users may prefer to switch to `branch = "something-else"` if their repository has both a `master` and default other branch.

In general the warnings in this PR aren't great. I've attempted to at least get the bare minimum warnings emitted but honestly it's pretty difficult to get much fancier than that. This behavior of unifying dependencies and fetching is very deep within Cargo and is (at least from what I was doing) difficult to connect to "go change this line in the manifest". Much of this is that Cargo isn't preserving that kind of information to assist in debugging since this has come up rarely as well.

Reply:
I've thought about "or alternatively place `branch = \"<main>\"` if you want to depend on the non-master main branch of a repository." or something like that, but it's very wordy and the warning is probably already a lot of text to sift through.